streamer/src/nonblocking/quic.rs

```diff
@@ -519,10 +558,72 @@ async fn setup_connection(
     }
 }
 
+async fn packet_batch_sender(
+    packet_sender: Sender<PacketBatch>,
+    packet_receiver: AsyncReceiver<PacketAccumulator>,
+    exit: Arc<AtomicBool>,
+    stats: Arc<StreamStats>,
+) {
+    trace!("enter packet_batch_sender");
+    loop {
+        let mut packet_batch = PacketBatch::with_capacity(PACKET_BATCH_SIZE);
+
+        stats
+            .total_packets_allocated
+            .fetch_add(PACKET_BATCH_SIZE, Ordering::Relaxed);
+
+        let mut last_sent = Instant::now();
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                return;
+            }
+            let elapsed = last_sent.elapsed();
+            if packet_batch.len() >= PACKET_BATCH_SIZE
+                || (!packet_batch.is_empty() && elapsed.as_millis() >= COALESCE_MS)
+            {
+                let len = packet_batch.len();
+                if let Err(e) = packet_sender.send(packet_batch) {
+                    stats
+                        .total_packet_batch_send_err
+                        .fetch_add(1, Ordering::Relaxed);
+                    trace!("Send error: {}", e);
+                } else {
+                    stats
+                        .total_packet_batches_sent
+                        .fetch_add(1, Ordering::Relaxed);
+                    trace!("Sent {} packet batch", len);
+                }
+                break;
+            }
+
+            let res = packet_receiver.try_recv();
+            if let Ok(packet_accumulator) = res {
+                unsafe {
+                    packet_batch.set_len(packet_batch.len() + 1);
+                }
+
+                let i = packet_batch.len() - 1;
+                *packet_batch[i].meta_mut() = packet_accumulator.meta;
+                for packet_bytes in packet_accumulator.bytes {
+                    packet_batch[i].buffer_mut()
+                        [packet_bytes.offset as usize..packet_bytes.end_of_chunk]
+                        .copy_from_slice(&packet_bytes.bytes);
+                }
+
+                if packet_batch.len() == 1 {
+                    last_sent = Instant::now();
```

Comment:
Why not set this really after the last_sent in the send block?

Reply:
This really should be named batch_started or something - will change the name. But basically, the idea is that it stores the time when we placed the first packet into the packet batch. This is the "start time" for the entire batch - and we want to send the batch in a short time (e.g., ~5ms) after the batch was started. We don't want to set this after we send the last batch, since it may be some time until we get a another packet to start another batch. Again, the variable name is bad, I will change it.